### DISCUSSION TITLE
How to show error again after script abort

### DISCUSSION THREAD

**Discussion by astirfreak** [micropython/micropython#11119] [General]
Hello,
I have a larger micropython script running on an ESP32. After some hours or days it seems, the script cancels because of an exception. I can login via web repl and get a working prompt. A new import of the script has no effect, so I'm sure, a reboot es not the cause of the problem. 
After a reboot, an import of the script always runs the script, which doesn't work after an exception.
Is there a way to repeat the last error output of a script?
Currently I have no idea which module of my script throws the exception. 
I can put a serial data logger (which I currently doesn't have) on serial interface of the ESP-module to get the root cause. But es there a smarter approach?

**Comment by davefes** [comment]
Not sure it is smarter ... but what I do in the exception block is write the resultant error to a errors.txt which is stored in flash.  Then you can re-visit that file on the next boot.

**Reply by astirfreak** [reply]
Do you have a code example, how to do that? I will give that a try,

**Reply by davefes** [reply]
```py
        try:
            sta_if.connect(config.hotspot, config.password)
        except OSError as error:
            try:
                with open('errors.txt', 'a') as outfile:
                    outfile.write(str(error) + '\n')
            except OSError:
                pass
```

**Reply by astirfreak** [reply]
Stupid question:
Does OSError catch really all exceptions?

**Reply by davefes** [reply]
Actually, a good question.  I am sure I read somewhere that all Micropython errors could be trapped this way.  I believe that you can be more specific by saying something like:
`except ValueError:`
Try it, I have trapped a lot of errors this way.

Google took me here:
https://forums.raspberrypi.com/viewtopic.php?t=342966

And here:
https://docs.micropython.org/en/v1.19/library/builtins.html

**Reply by GitHubsSilverBullet** [reply]
```py
try:
    try:
        a = 1 / 0
    except ZeroDivisionError as e:
        print(f'Inner:{type(e)}')
    a = 1 / 0
except Exception as e:
    print(f'Outer:{type(e)}')

# Inner:<class 'ZeroDivisionError'>
# Outer:<class 'ZeroDivisionError'>
```
The first error is caught specifically, the second one generally.

**Reply by astirfreak** [reply]
Ok, I did a first try with this code:
```
try:
	a=1/0
except Exception as e:
	print(str(e))
```
output:
```
C:\test>python test.py
division by zero
```
and now without try/except:
```
Traceback (most recent call last):
  File "test.py", line 4, in <module>
    a=1/0
ZeroDivisionError: division by zero
```

So the standard console output delivers much more information, where the exception occured compared to my exception handler. How to get this detailed information to put it in the log-file too?

**Comment by GitHubsSilverBullet** [comment]
Try this:
```py
import io
import sys

def bad_func():
    return 1 / 0

def good_func():
    try:
        return 1 / 0
    except ZeroDivisionError as e:
        print('-'*50)
        print(f'good_func:{type(e)}')
        print(f'good_func:{__name__}')
        output = io.StringIO()
        sys.print_exception(e, output)
        print(output.getvalue())
        print('-'*50)
    return None

try:
    good_func()
    bad_func()
except Exception as e:
    print('='*50)
    print(f'Outer:{type(e)}')
    print(f'Outer:{__name__}')
    output = io.StringIO()
    sys.print_exception(e, output)
    print(output.getvalue())
    print('='*50)
```

**Reply by astirfreak** [reply]
sys.print_exception seems to be a good approach.
But how to put this output into a logfile?

```python
sys.print_exception(e, file="errors.txt")
```
generates the following output:
```python
>>> import test                                                                                 
Traceback (most recent call last):                                                              
  File "<stdin>", line 1, in <module>                                                           
  File "test.py", line 16, in <module>                                                          
TypeError: function doesn't take keyword arguments
```

**Comment by GitHubsSilverBullet** [comment]
```py
output = io.StringIO()  # prepare 
sys.print_exception(e, output)  # print info into memory
s = output.getvalue() # retrieve the whole string
# do whatever you like to do with that string
```

If you just want to log the raw data:
```py
try:
    1 / 0
except Exception as e:
    with open('logfile', 'a') as fh:
        sys.print_exception(e, fh)
```

**Comment by astirfreak** [comment]
Thank you all for your support:

This one seems to work for me:
```python
try:
    main()
except Exception as e:
	t=utime.localtime()
	print("Exception")
	errStr="\n===== Exception at "+str(t[3])+":"+str(t[4])+" "+str(t[2])+"."+str(t[1])+"."+str(t[0])+" =====\n"	
	with open('errors.txt', 'a') as outfile:
		outfile.write(errStr)
		sys.print_exception(e, outfile)
	outfile.close()
	print("errors.txt written successfully")
	sys.exit()
```

**Reply by GitHubsSilverBullet** [reply]
```py
with open('errors.txt', 'a') as outfile:
	outfile.write(errStr)
	sys.print_exception(e, outfile)
	outfile.close()
```
The ›close‹ is superfluous because the ›with‹ clause takes already care of closing the file as soon as the block is left.

**Reply by astirfreak** [reply]
In a first try, the script didn't generate the file. So I added the close-statement and did some other changes at the same step. It then the works. Maybe that was a fluke, but I didn't do any further research after the script was working.

**Comment by astirfreak** [comment]
By the way, I asked chatGPT (in German, translated by Deepl):

> I have a Micropython script on an ESP32, with many modules, which keeps aborting. Since I don't have a REPL connection to the ESP32 at the time of the abort, I don't know why the script aborted. Is there a way to have the REPL output of the exception reissued after establishing a REPL connection?

> Yes, there is a way to have the REPL output of the exception reissued after a REPL connection has been established.
If the script crashes due to an exception, an error message is normally output to the console. However, if you do not have a REPL connection at the time of the crash, you will not be able to see the error message.
However, if you establish the REPL connection later, you can retrieve the latest error messages on the ESP32 by using the sys.print_exception() method.
Here is a sample code that retrieves and prints the last error messages as soon as a REPL connection is established:

```python
import sys

# diese Methode wird aufgerufen, sobald eine REPL-Verbindung hergestellt wird
def on_repl():
    # prüfen, ob eine Exception aufgetreten ist
    if sys.exc_info()[0] is not None:
        # die letzte Exception ausgeben
        sys.print_exception(*sys.exc_info())

# diese Methode registriert die "on_repl"-Methode als Callback
def register_on_repl_callback():
    import builtins
    if hasattr(builtins, "__REPL__"):
        builtins.__REPL__.register_repl_callback(on_repl)

# Callback-Registrierung aufrufen
register_on_repl_callback()
```

> You can place this code at the beginning of your script to ensure that the on_repl() method is called as soon as a REPL connection is established. If an exception has occurred, the last error message will be issued.

I didn't try it because I didn't understand it.

**Comment by sosi-deadeye** [comment]
Where do the German comments come from?
Did the AI write that or is that an artifact from BigData?

`Exception` catch mostly everything.
My try:
```python

import sys
import time
import uasyncio as asyncio

from random import choice
from uasyncio import TimeoutError


ISO_FMT = "{}-{:02d}-{:02d}T{:02d}:{:02d}:{:02d}"


def random_exception():
    exceptions = (
        ValueError("You can't do this!"),
        TypeError("I don't understand your operation."),
        RuntimeError("Something which happens during runtime"),
        TimeoutError("Timeout"),
        KeyError("Key does not exist"),
        IndexError("Index does not exist.")
    )
    raise choice(exceptions)


async def worker():
    return 1 / 0


async def async_exception():
    await asyncio.create_task(worker())


def main():
    if choice([True, False]):
        print("Sync code")
        random_exception()
    else:
        print("Async code")
        asyncio.run(async_exception())


def exc_catcher(debug=False):
    if not debug:
        return main()

    try:
        main()
    except Exception as e:
        with open("error.txt", "a") as fd:
            fd.write(ISO_FMT.format(*time.localtime()[:6]))
            fd.write("\n")
            sys.print_exception(e, fd)
            fd.write("\n")


if __name__ == "__main__":
    exc_catcher(debug=True)
    print("Program End")
```

**Reply by astirfreak** [reply]
> Where do the German comments come from?
Did the AI write that or is that an artifact from BigData?

The complete conversation was in german. I translated it to english with deepl.
The comments came from the AI, which I didn't translate.

The solution, I posted above, is active now. 
... waiting for the next exception ... (which didn't come last night)
